File: src/main/scala/org/zezutom/learningspark/scala/WordCountAndJoin.scala
package org.zezutom.learningspark.scala

import org.apache.spark.rdd.RDD

object WordCountAndJoin {

  def wc_join(keyword: String, rdd1: RDD[String], rdd2: RDD[String]): Array[(String, (Int, Int))] = {
    val wc_out1 = wc(keyword, rdd1)
    val wc_out2 = wc(keyword, rdd2)
    wc_out1.join(wc_out2).collect()
  }

  private def wc(keyword: String, rdd: RDD[String]): RDD[(String, Int)] =
    rdd
      .flatMap(line => line.split("\\s"))                     // Split on any white character
      .map(_.stripSuffix(",").stripSuffix(".").toLowerCase)   // Remove punctuation and make all words lowercase
      .filter(_.equals(keyword.toLowerCase))
      .map(word => (word, 1)).reduceByKey(_ + _)
  
}

File: src/main/java/org/zezutom/learningspark/WordCountAndJoin.java
package org.zezutom.learningspark;

import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import scala.Tuple2;

import java.util.Arrays;
import java.util.List;

public class WordCountAndJoin {

    public static List<Tuple2<String, Tuple2<Integer, Integer>>> wc_join(String keyword, JavaRDD<String> rdd1, JavaRDD<String> rdd2) {
        JavaPairRDD<String, Integer> wc_out1 = wc(keyword, rdd1);
        JavaPairRDD<String, Integer> wc_out2 = wc(keyword, rdd2);
        return wc_out1.join(wc_out2).collect();
    }

    private static JavaPairRDD<String, Integer> wc(String keyword, JavaRDD<String> rdd) {
        return rdd
                .flatMap(s -> Arrays.asList(s.split("\\s")))    // Split on any white character
                .map(s -> s.replaceAll(",", "")
                        .replaceAll("\\.", "").toLowerCase())   // Remove punctuation and make all words lowercase
                .filter(s -> s.equals(keyword.toLowerCase()))
                .mapToPair(s -> new Tuple2<>(s, 1))
                .reduceByKey((i1, i2) -> i1 + i2);
    }
}
